observIQ/bindplane-op-action
# feat: Initial Action

Initial action has the following features
- Apply destinations from a given file path
- Apply configurations from a given file path
- Write raw otel configurations back to the repo

The repo contains Ci configuration for testing against a bindplane instance (local docker).
- Shellcheck for the script
- Testing against 1.37 and 1.40 bindplane releases. This list can be expanded as needed.

## Testing

I have an example repo [here](https://github.com/jsirianni/bindplane-op-action-test/actions/runs/7630650838), but you should try it on your own.

You can test with the following steps

1. Create a new bindplane op account within our dev or stage environments
2. Create an API key
3. Create one or more configurations with one or more destinations
4. Create a new repo under your personal github, give it any name. You can delete it after.
  - Create repo actions secret `BINDPLANE_REMOTE_URL` with value `https://<bindplane dev or stage>.com`
  - Create repo actions secret `BINDPLANE_SECRET_KEY`
6. Export the resources to "seed" your repo
```bash
bindplane get destination -o yaml --export > destinations.yaml
bindplane get configuration -o yaml --export > configurations.yaml
```
7. Create your workflow file with the following contents
```yaml
name: bindplane
on:
  push:
    branches:
      - main
permissions:
  contents: write
concurrency:
  group: ${{ github.head_ref || github.ref_name }}
  cancel-in-progress: false
jobs:
  bindplane:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          fetch-depth: 0

     # Pointing to the dev branch because main is not ready yet
      - uses: observIQ/bindplane-op-action@dev
        with:
          bindplane_remote_url: ${{ secrets.BINDPLANE_REMOTE_URL }}
          bindplane_api_key: ${{ secrets.BINDPLANE_SECRET_KEY }}
          target_branch: main
          destination_path: destination.yaml
          configuration_path: configuration.yaml
          enable_otel_config_write_back: true
          configuration_output_dir: otel/
          token: ${{ secrets.GITHUB_TOKEN }}
```
8. Commit everything to your repo.

You should expect actions to kick off. You should also expect to see a new commit from the action containing your configurations at the directory `otel/`.

If you change your configurations / destinations, you should expect them to be applied. You should also expect to see a pending rollout in the ui.

edit: We added an `enable_auto_rollout` option. If you include an additional `with` param, `enable_auto_rollout: true`, you should expect the configuration to automatically rollout.
